File: Basic_Calculator/python/basic_calculator.py
def calculate(s):
    """
    Evaluate a mathematical expression string without using eval().
    
    Args:
        s (str): A string representing a valid mathematical expression
        
    Returns:
        int: The result of the evaluation
    """
    # need a stack
    # get rid of spaces first
    if not s:
        return 0
    sflat = s.replace(' ', '')


    #(1+(4+5+2)-3)
    # what we want is to get the evaluate numbers that are inside parentheses first.
    # I am debating about how to process nested parentheses.  
    # I have a feeling I need recursion. I am wanting to make a global stack where 
    # I put the string after I have remove all the spaces. 
    # Then I am going to process left to right

    def calc(s: str, index: int) -> (int, int):
        stack = []
        num = 0
        sign = 1
        localTotal = 0

        while index < len(s):
            char = s[index]
            if char.isdigit():
                num = num * 10 + int(char)
            elif char in '+-':
                stack.append(sign * num)
                num = 0
                if char == '+':
                    sign = 1
                else:
                    sign = -1
            elif char == '(':
                num, index = calc(s, index + 1)
            elif char == ')':
                stack.append(sign * num)
                for elem in stack:
                    localTotal += int(elem)
                return localTotal, index
            index += 1
        stack.append(sign * num)
        localTotal = 0
        for elem in stack:
            localTotal += int(elem)
        return localTotal, index
    
    total, empty = calc(sflat, 0)
    return total

File: Basic_Calculator/cpp/basic_calculator.cpp
#include <string>
#include <stack>
using namespace std;

int calculate(string s) {
    stack<int> stk;
    int num = 0;
    int sign = 1;  // 1 for positive, -1 for negative
    int result = 0;
    
    for (char c : s) {
        if (isdigit(c)) {
            // Build the current number
            num = num * 10 + (c - '0');
        } else if (c == '+') {
            // Add the current number to result and reset
            result += sign * num;
            num = 0;
            sign = 1;
        } else if (c == '-') {
            // Add the current number to result and set negative sign
            result += sign * num;
            num = 0;
            sign = -1;
        } else if (c == '(') {
            // Push current result and sign to stack, reset for new scope
            stk.push(result);
            stk.push(sign);
            result = 0;
            sign = 1;
        } else if (c == ')') {
            // Add current number, then pop sign and previous result
            result += sign * num;
            num = 0;
            result *= stk.top(); stk.pop();  // multiply by sign before parentheses
            result += stk.top(); stk.pop();  // add result before parentheses
        }
        // Skip spaces automatically
    }
    
    // Add the last number
    result += sign * num;
    return result;
} 